students/JasneetChandok/Lesson04/mailroom2.py

```diff
@@ -0,0 +1,172 @@
+    print(
+        '''
+              Dear {}
+
+              Thank you for your very kind donation of ${:.2f}.
+              It will be put to very good use.
+
+                             Sincerely,
+                                -The Team
+        '''.format(donor_selection, donor_donation))
+
+
+# Function for sending a thanking you note
+def thankyou_email():
+    donor_category = input("""Specify if new or pick existing donor from the list
+        New Donor: Enter 'new'
+        Existing Donor: Enter 'list'
+        >>>""")
+    donor_category = donor_category.strip().lower()
+    print("\nYou selected:", donor_category, "\n")
+    if donor_category == "new":
+        new_donor()
+    elif donor_category == "list":
+        existing_donor()
+    else:
+        print("Select one of the two category!!")
+
+
+# Function for sorting report by amount
+def sort_key(item):
+    return item[1]
+
+
+# Function for generating reporting
+def generate_report():
+    report_rows = []
+    for (name, gifts) in donors.values():
+        total_gifts = sum(gifts)
+        num_gifts = len(gifts)
+        avg_gift = total_gifts / num_gifts
+        report_rows.append((name, total_gifts, num_gifts, avg_gift))
+
+    # sort the report data
+    report_rows.sort(key=sort_key)
+    report = []
+    report.append("{:25s} | {:11s} | {:9s} | {:12s}".format("Donor Name",
+                                                            "Total Given",
+                                                            "Num Gifts",
+                                                            "Average Gift"))
+    report.append("-" * 66)
+    for row in report_rows:
+        report.append("{:25s}   ${:10.2f}   {:9d}   ${:11.2f}".format(*row))
+    return "\n".join(report)
+
+
+def gen_letter(donor):
+    return ('''Dear {0:s},
+
+          Thank you for your very kind donation of ${1:.2f}.
+          It will be put to very good use.
+
+                         Sincerely,
+                            -The Team
+          '''.format(donor[0], donor[1][-1]))
+
+
+def save_letters_to_disk():
+    for donor in donors.values():
+        letter = gen_letter(donor)
+        filename = donor[0].replace(" ", "_") + ".txt"
+        print("writing letter to:", donor[0])
+        open(filename, 'w').write(letter)
+
+
+# Function for quitting the propram
+def quit_program():
+    print("Program Quit!!")
+    sys.exit(0)
+
+
+def main_menu():
+    prompt = ('''
+                        Choose an action:
+
+                        (1) - Send a Thank You
+                        (2) - Create a Report
+                        (3) - Send letters to everyone
+                        (4) - Quit
+
+                        > ''')
+
+    selection_dict = {"1": thankyou_email,
+                      "2": generate_report,
+                      "3": save_letters_to_disk,
+                      "4": quit_program}
+
+    run_menu(prompt, selection_dict)
+
+
+def run_menu(prompt, selection_dict):

```

Comment:
I'm not able to get this working when I try to print the report. Try running it again and see if you're able to get this behavior to work.